cylc/flow/scripts/lint.py

```diff
@@ -0,0 +1,392 @@
+#!/usr/bin/env python3
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Look through one or more folders for Cylc 7 syntax ``suite*.rc`` files
+and search for syntax which may be problematic at Cylc 8.
+
+Can be run either as a linter or "in place" (``-i``), leaving comments
+in files. If used in the "in place" mode it is recommended that you ensure
+that you have recorded the state of your workflow in a version control
+system before starting.
+"""
+from colorama import Fore
+from optparse import Values
+from pathlib import Path
+import re
+from typing import Generator
+
+from cylc.flow import LOG
+from cylc.flow.option_parsers import (
+    CylcOptionParser as COP,
+)
+from cylc.flow.terminal import cli_function
+
+STYLE_GUIDE = (
+    'https://cylc.github.io/cylc-doc/latest/html/workflow-design-guide/'
+    'style-guide.html#'
+)
+URL_STUB = "https://cylc.github.io/cylc-doc/latest/html/7-to-8/"
+SECTION1 = r'\[{}\]'
+SECTION2 = r'\[\[{}\]\]'
+SECTION3 = r'\[\[\[{}\]\]\]'
+FILEGLOBS = ['*.rc', '*.cylc']
+CHECKS = {
+    '7-to-8': {
+        re.compile(SECTION1.format('vizualization')): {
+            'short': 'section ``[vizualization]`` has been removed.',
+            'url': 'summary.html#new-web-and-terminal-uis'
+        },
+        re.compile(SECTION1.format('cylc')): {
+            'short': 'section ``[cylc]`` is now called ``[scheduler]``.',
+            'url': 'summary.html#terminology'
+        },
+        re.compile(SECTION2.format('authentication')): {
+            'short': '``[cylc][authentication]`` is now obsolete.',
+            'url': ''
+        },
+        re.compile(r'include at start-up\s*?='): {
```

Comment:
I think these should include the caret so as to only match the exact phrase and not things that end with the phrase
```suggestion
        re.compile(r'^\s*include at start-up\s*='): {
```
(There is no need for a `?` after `x*` - `*` means match 0 or more times)